✨ Feat: 상품 주문 API 추가

m1naworld opened the pull request:
<!-- 
체크하려면 괄호 안에 "x"를 입력하세요. 
각 규칙은 Convention 문서에 있습니다.
PR 제목에 쓰는 prefix는 다음과 같습니다.

✨ Feat : 새로운 기능
🔨️ Refactor : 코드 리팩토링
🐎 Perf : 성능을 향상
🐛 Fix : 버그를 고칠 때
🧪 Test : 테스트 코드
🚜 Rename : 파일 이름 변경 혹은 구조를 변경
🚀 Release : 배포 / 개발 작업과 관련된 모든 것
🔥 Remove : 코드 또는 파일 제거
📚 Docs : 문서
📝 Chore : 사소한 코드 또는 언어를 변경 기타 변경사항 (빌드 스크립트 수정, 패키지 매니징 설정 등)

-->

## PR 타입
<!-- 어떤 유형의 PR인지 체크해주세요. -->

<!-- 체크하려면 괄호 안에 "x"를 입력하세요. -->
- [x] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Performance Improvement
- [ ] Bugfix
- [x] Test Code
- [ ] Code style update (formatting, local variables)
- [ ] Other... Please describe:

## 개요 
- 상품 주문 기능 구현


## 작업 및 변경 사항 
- 상품 주문 기능 구현 

<br/>

### 공유 하고 싶은 내용 및 참고 자료

❗️ 우선 제안 사항입니다. 
common > ResponseDto를 보내주는 데이터의 유무에 따라 Respons, DataResponse로 나눴음 하는 바람입니다. 
<img width="526" alt="스크린샷 2023-01-07 오후 3 13 06" src="https://user-images.githubusercontent.com/85235063/211134105-ca1dde5d-2e13-4aab-9bf9-c317e9b6e62c.png">
저는 ResponseDto에 @JsonInclude(JsonInclude.Include.NON_NULL) 설정을 해주었지만, null을 보내는게 이질적으로 느껴집니다. 
<br/>

<img width="796" alt="스크린샷 2023-01-07 오후 3 13 38" src="https://user-images.githubusercontent.com/85235063/211134123-7c9ce7ee-34db-4459-9de8-ccf520e5ee8d.png">
또한 이러한 경우 컨트롤러 부분에서 제네릭을 통해 데이터 타입이 한눈에 보이지않아 현재 데이터를 함께 반환하는지, 반환하지 않는 경우인지 한눈에 파악하기 힘듭니다.


#### dto를 나누던가 혹은 데이터가 보내지는 Dto라면 제네릭에 데이터 타입을 꼭 표기하여 null일 경우와 구분할 수 있도록 컨벤션을 정하는 것도 좋아보입니다!  


close #5

m1naworld committed:
[docs] issue 및 PR 템플릿 파일 수정

m1naworld committed:
Merge remote-tracking branch 'origin/main' into main

m1naworld committed:
[docs] PR 템플릿 파일 수정

m1naworld committed:
Merge remote-tracking branch 'origin/develop' into develop

m1naworld committed:
[refactor] Orders entity 수정 및 OrderedProduct entity 삭제

DB 구조 Orders & OrderedProduct 통합 → entity, repository 수정 및 삭제

Related to: #5

m1naworld committed:
Merge remote-tracking branch 'origin/develop' into develop

m1naworld committed:
[refactor] Orders entity 수정

컬럼명 amount → quantity 변경 및 createdAt 자동 생성 설정

Related to: #5

m1naworld committed:
[feat] 상품 주문 기능 추가

상품 주문 기능 추가

Related to: #5

m1naworld committed:
[refactor] ResponseDto 제네릭 표기

Related to: #5

hyunmuam commented on src/main/java/com/hanbang/e/order/controller/OrderController.java:
```diff
@@ -0,0 +1,33 @@
+	private final OrderService orderService;
+
+	@PostMapping("")
+	public ResponseEntity<ResponseDto<?>> doOrder(@RequestParam Long productId, @RequestBody OrderReq orderReq) {
```
`ResponseEntity<ResponseDto<?>>`를 그냥 `ResponseEntity<?>`로 했어도 될텐데 그렇게 하지 않으신 이유가 있을까요?

m1naworld replied:
@tjvm0877 

우선 현재 API 당 반환되는 타입은 일정 합니다. 그래서 `와일드 카드(?)`를 사용해야 하는 이유를 찾지 못했습니다.
또한 오히려 유지 보수 시 등을 고려하면 와일드 카드를 통해 확장성을 미리 열어두는 것보다는 타입을 지정하는 측면이 가시성, 혹시 모를 에러 측면에 더 좋다고 생각했기 때문입니다. 
또한 컨트롤러는 클라이언트와 통신하는 부분으로 반환되는 것이 무엇인지를 한눈에 파악할 수 있는 것이 중요하다고 생각했습니다.

재현님의 생각은 어떠신가요?

hyunmuam commented on src/main/java/com/hanbang/e/order/service/OrderService.java:
```diff
@@ -0,0 +1,50 @@
+
+		int orderQuantity = orderReq.getQuantity();
+
+		product.updateSales(orderQuantity);
+
```
판매량을 증가시키는 로직은 있는데 재고량을 빼는 로직은 존재하지 않는 것 같습니다.
또한 주문량이 판매량 보다 많았을 때 잡아주는 부분이 없는 것 같습니다.

m1naworld replied:
@tjvm0877 우선, 재고량은 순수 고정으로 두고 판매량을 통해 현재 남은 재고를 생각하면 되겠다고 판단했습니다. 그 이유는 재고량을 기준으로 어느정도를 판매했는지를 판단할 수 있기에 재고량은 고정으로 두는 것이 옳다고 판단하였습니다. 이부분에 대해서도 얘기를 나눠보면 좋을 것 같습니다! 
또한 주문량이 판매량보다 많았을 때를 놓치다니 정말 아차차,, 네요 🤣 바로 추가하도록 하겠습니다 👀 매의 눈 최곱니다!

hyunmuam commented on src/main/java/com/hanbang/e/product/entity/Product.java:
```diff
@@ -53,4 +53,8 @@ public Product(String productName, Long price, String img, int stock, int sales,
 		this.brand = brand;
 	}
+
+	public void updateSales(int quantity){
+		this.sales += quantity;
```
메소드 이름을 조금 더 생각해봐도 될 것 같습니다. 위에서 말했듯이 판매량은 증가하지만  `updataSales`은 판매량만 증가시켜줬습니다. 우리가 "**판매했다**." 라는 행위라는 관점으로 접근하면 판매량을 올리고, 재고는 감소시키는 메소드로 묶어 줄 수 있을 것 같습니다. 매소드 이름을 그에 맞게 바꿔도 될 것 같고요.

m1naworld replied:
@tjvm0877 이 부분에 대해서는 위에 논의를 거친 후에 다시 한번 생각 해 보겠습니다 🫡

hyunmuam reviewed (changes requested):
1. 같이 PR에 들어온 템플릿 관련된 파일은 삭제해주세요. 가능하다면 커밋 기록까지 삭제 부탁드립니다.
2. 코드 확인했습니다. 코멘트 확인 부탁드립니다.
3. 제안 사항 확인했습니다. 좋은 의견인 것 같아요. 제가 담당하는 부분에 적용해도 될 것 같습니다. Issue올려주시고 어떻게 바꾸면 좋을지 딱 정해버리시죠! 의논이 끝나면 노션에 있는 컨벤션에도 추가해도 될 것 같습니다.

m1naworld committed:
[refactor] 상품 주문시 판매량, 재고량 변경 메소드

1. 재고량도 차감하는 로직으로 변경
2. 코드 가독성을 위한 메소드명 변경

Related to: #5

m1naworld committed:
[refactor] Dto → Entity 변환 컨벤션 변경에 따른 코드 수정

1. 변환 메소드 패턴 변경: 팩토리 메소드 + 빌더를 통한 구현에서 빌더만을 활용
2. 패턴이 변경됨에 따라 메소드 명 및 위치 변경

Related to: #5

m1naworld committed:
[refactor] 상품 주문 로직 조건 추가

1. 현재 재고량보다 주문 수량이 많을 경우 조건 처리
2. 주문 수량이 0으로 들어올 경우 @Valid + ValidExceptionHandler를 통해 예외처리

Related to: #5

m1naworld committed:
[refactor] Orders 생성자 파라미터 값 변경

1. 생성자 파라미터에서 totalPrice 제거
2. 그에 따른 dto → entity 변환 메소드 수정

Related to: #5

m1naworld committed:
[test] 상품 주문 Service 계층 테스트 코드 작성

1. 테스트 코드에 필요한 라이브러리 의존성 추가
2. 상품 주문 기능 서비스 계층 테스트 코드 추가

Related to: #5

hyunmuam commented on src/main/java/com/hanbang/e/order/controller/OrderController.java:
```diff
@@ -0,0 +1,36 @@
+
+	@PostMapping("")
+	public ResponseEntity<ResponseDto<?>> doOrder(@RequestParam Long productId, @Valid @RequestBody OrderReq orderReq) {
+		Long memberId = 1L; // 임시용
```
이부분은 지워주셔도 될 것 같습니다.

m1naworld replied:
Long memberId 말씀이시죠?! 이부분은 나중에 jwt가 구현이 된 후 수정할 때 한눈에 파악할 수 있도록 일부로 남겨 놓은건데요! 저는 계속 남겨두고 싶은 부분입니다. 어떻게 생각하시나요?! 🥹..

hyunmuam replied:
복잡한 로직도 아니고 한줄이라서 남겨도 딱히 큰 의미는 없을 것 같아요.

hyunmuam commented on src/main/java/com/hanbang/e/common/exception/GlobalExceptionHandler.java:
```diff
@@ -21,4 +24,15 @@ public ResponseEntity IllegalArgumentExceptionHandler(Exception e) {
+		String[] messageParsing = e.getMessage().split("\\[|\\]");
+		String message = messageParsing[messageParsing.length - 3];
+
+		ResponseDto response = new ResponseDto("fail", message, null);
```
클라이언트에서 무엇이 문제인지 알 수 있도록 message에 추가해주면 좋을 것 같습니다.

m1naworld replied:
클라이언트에서 확인할 수 있는 message는 `@Min(value = 1, message = "주문 수량을 선택해주세요.")` 이를 통해 '주문 수량을 선택해주세요'가 들어가게 설정해두었습니다! :)

hyunmuam replied:
아 그러네요 좋습니다.

hyunmuam commented on build.gradle:
```diff
@@ -24,8 +24,15 @@ dependencies {
+
     testImplementation 'org.springframework.boot:spring-boot-starter-test'
     testImplementation("org.assertj:assertj-core:3.23.1")
+    testCompileOnly 'org.projectlombok:lombok'
+    testAnnotationProcessor 'org.projectlombok:lombok'
```
테스트 코드를 보았을 때 lombok어노테이션이 사용되는 것 같지 않는거 같은데. 필요없는 Dependency는 지워주셔도 될 것 같습니다.

m1naworld replied:
꼼꼼한 리뷰 정말 감사합니당!! 🫡

hyunmuam commented on src/test/java/com/hanbang/e/order/service/OrderServiceTest.java:
```diff
@@ -0,0 +1,213 @@
+package com.hanbang.e.order.service;
+
+import static org.assertj.core.api.Assertions.*;
+import static org.mockito.Mockito.*;
```
와일드 카드는 해제해주시고 simgle import해주세요~ 다른 코드들에서도 다시한번 확인 부탁드립니다.

m1naworld replied:
캠퍼스 핵데이 컨벤션을 보면 static import에는 와일드 카드가 허용된다고 합니다! 그래서 이렇게 적용하였는데, 이 부분도 함께 이야기해보면 좋을 것 같습니다!

hyunmuam replied:
오~ 그러네요. 제가 확인을 못했습니다. 그냥 나둬도 될 것 같습니다.

hyunmuam reviewed (changes requested):
리뷰 확인 부탁드립니다. 이전 코멘트에 대해서 해결되었다면 ResolveConversation을 눌러 처리했음을 볼 수 있음 좋겠습니다.

hyunmuam committed:
[test] 상품 주문 Controller 계층 테스트 코드 작성

1. 테스트 코드에 필요한 라이브러리 의존성 추가
2. 상품 주문 기능 컨트롤러 계층 테스트 코드 추가

Related to: #5

m1naworld committed:
[chore] 사용하지 않는 라이브러리 삭제

testCompileOnly 'org.projectlombok:lombok'
testAnnotationProcessor 'org.projectlombok:lombok' 제거
